File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head == null || head.next == null)
            return head;
        int i=0;
        ListNode temp = head;
        while(temp != null && i < k-1){
            temp = temp.next;
            i++;
        }
        if(temp == null)
            return head;
        ListNode nextHead = reverseKGroup(temp.next, k);
        temp.next = null;
        ListNode h = head;
        ListNode newHead = reverse(head);
        h.next = nextHead;
        return newHead;
        
    }
    
    public ListNode reverse(ListNode head){
        if(head == null || head.next == null)
            return head;
        ListNode next = reverse(head.next);
        head.next.next = head;
        head.next = null;
        return next;
    }
}

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if k == 1 and head == None:
            return head
        temp = head
        prev = None
        i = 0
        while i < k and temp != None:
            temp = temp.next
            i += 1
        if i < k:
            return head
        temp = head
        i = 0
        while i < k  and temp != None:
            nex = temp.next
            temp.next = prev
            prev = temp
            temp = nex
            i += 1
        head.next = self.reverseKGroup(temp, k)
        return prev
            